### PR TITLE
Add a visibility suggestion in private-in-public errors

### LINKED ISSUE #112284: Suggest code correction for E0446 - private type in public interface
### Code

```Rust
struct MyStruct;
pub fn my_test(_v: &MyStruct) {}
fn main() {}
```

### Current output

```Shell
error[E0446]: private type `MyStruct` in public interface
 --> src/bin/main.rs:2:1
  |
1 | struct MyStruct;
  | --------------- `MyStruct` declared as private
2 | pub fn my_test(_v: &MyStruct) {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't leak private type
```

### Desired output

```Shell
error[E0446]: private type `MyStruct` in public interface
 --> src/bin/main.rs:2:1
  |
1 | struct MyStruct;
  | --------------- `MyStruct` declared as private
2 | pub fn my_test(_v: &MyStruct) {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't leak private type

Help:
* Either remove the `pub` in `pub fn my_test`
* Or add `pub` to `struct MyStruct;`
```

### Rationale and extra context

Novice developers might be confused by the current wording of "leaking private type", and this seems like an easy help message fix.

### Other cases

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by nyurik**
This modifies the original #112540 by @aryan-debug to fix #112284

This PR introduces a new field `vis_sugg` to suggest the proper visibility for the error.

TODO:   What should the `TODO` values be here?  Note that not a single test anywhere has used these `???TODO???` values, so it is unclear how they can be generated.  Maybe this whole match should be simplified to map `Restricted(_) => "pub(crate)"` ?

```rust
let vis_sugg = match self.required_visibility {
    ty::Visibility::Public => "pub",
    ty::Visibility::Restricted(vis_def_id) => {
        if vis_def_id == self.tcx.parent_module(hir_id) {
            "???TODO???"
        } else if vis_def_id.is_top_level_module() {
            "pub(crate)"
        } else {
            "???TODO???"
        }
    }
};
```

### CODE REVIEW

**Comment by WaffleLapkin** [comment]
r? compiler-errors

**Comment by petrochenkov** [comment]
I don't think this change is still relevant after https://github.com/rust-lang/rust/pull/113126.

**Comment by nyurik** [comment]
@petrochenkov thx for the heads up! I looked through that PR (even left a comment), but I'm not certain I understand if it will cover all cases?

I see [this change](https://github.com/rust-lang/rust/pull/113126/files#diff-9a9bf53a67d6c204294aa9ab4ec8fd1265306998e0eefc593f004e053513b823R1472) adds `self.in_assoc_ty` bool, so clearly the error will still exist if its an associated type? Is this planned to change as well? Should the suggestion I did in this PR be addapted to #113126?

```rust
if self.in_assoc_ty && !vis.is_at_least(self.required_visibility, self.tcx) {
```

**Comment by compiler-errors** [comment]
r? petrochenkov

**Comment by petrochenkov** [comment]
Blocked on https://github.com/rust-lang/rust/pull/113126.
@rustbot blocked

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/113126 has landed.
@rustbot author

**Comment by Dylan-DPC** [comment]
@nyurik any updates?

**Comment by nyurik** [comment]
@Dylan-DPC thx for the ping, i have updated the PR, but it still has some outstanding questions that need some feedback. Thx!

**Comment by nyurik** [comment]
@petrochenkov hi, would you be the right person to review this?  I would love to move forward, but it is a bit unclear. Plus @JohnCSimon was pinging the original author of this PR (before my refactoring). Hopefully the outstanding questions can be easily addressed. Thx!!

**Comment by petrochenkov** [comment]
>@petrochenkov hi, would you be the right person to review this?

Yes, but maybe a bit later, I'm busy this week.

File: compiler/rustc_privacy/src/lib.rs [hunk 1]
```diff
@@ -1475,12 +1475,31 @@ impl SearchInterfaceForPrivateItemsVisitor<'_> {
                 }
             };
 
+            // FIXME: this code was adapted from the above `vis_descr` computation,
+            //        but it's not clear if it's correct.
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:1479]: I think `fn vis_to_string` should be usable here.

File: compiler/rustc_privacy/src/errors.rs [hunk 2]
```diff
@@ -55,9 +55,12 @@ pub struct InPublicInterface<'a> {
     pub span: Span,
     pub vis_descr: &'static str,
     pub kind: &'a str,
+    pub vis_sugg: &'static str,
     pub descr: DiagnosticArgFromDisplay<'a>,
     #[label(privacy_visibility_label)]
     pub vis_span: Span,
+    #[suggestion(code = "", applicability = "maybe-incorrect")]
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/errors.rs:62]: The suggested visibility should probably go to the `code` here, rather than to the suggestion message.

**Comment by Dylan-DPC** [comment]
@nyurik any updates on replying to the changes requested in the review and implement them accordingly?

**Review by nyurik — COMMENTED** [review]
Thx for the ping!  I am still interested in this PR, but got a bit stuck on implementing it

File: compiler/rustc_privacy/src/lib.rs (hunk 1 above)
Comment (petrochenkov) [reply compiler/rustc_privacy/src/lib.rs:1438]: I meant `ty::Visibility::to_string`, not `vis_to_string` from rustc_ast_pretty, sorry.

`ty::Visibility::to_string` was also called `vis_to_string` when this comment was written, but it was renamed in https://github.com/rust-lang/rust/pull/115993.

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/113983#discussion_r1486418222 wasn't addressed, the suggestion spans are incorrect and CI is failing.
@rustbot author

**Comment by JohnCSimon** [comment]
@nyurik 

Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you are going to continue please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive

**Pull request closed without merging**